walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,55 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> eventTimesOnlyMandatory = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> eventTimesWithOptional = new ArrayList<TimeRange>();
+    for (Event event : events) {
+	  Set<String> intersection = new HashSet<String>(request.getAttendees());
+	  intersection.retainAll(event.getAttendees());
+      if (intersection.size() > 0) {
+        eventTimesOnlyMandatory.add(event.getWhen());
+        eventTimesWithOptional.add(event.getWhen());
+      } else {
+        intersection = new HashSet<String>(request.getOptionalAttendees());
+        intersection.retainAll(event.getAttendees());
+        if(intersection.size() > 0) {
+          eventTimesWithOptional.add(event.getWhen());
+        }
+      }
+    }
+    Collection<TimeRange> options =  findFreeTimes(eventTimesWithOptional, request.getDuration());
+    if (options.size() > 0 || request.getAttendees().size() == 0 || request.getOptionalAttendees().size() == 0) {
+        return options;
+    }
+    return findFreeTimes(eventTimesOnlyMandatory, request.getDuration());
+  }
+
+  private Collection<TimeRange> findFreeTimes(ArrayList<TimeRange> eventTimes, long duration) {
+    Collections.sort(eventTimes, TimeRange.ORDER_BY_START);
+    int start = TimeRange.START_OF_DAY;
+    ArrayList<TimeRange> options = new ArrayList<TimeRange>();
+    for (TimeRange eventTime : eventTimes) { 
+      if(eventTime.start() > start) {
+        if (eventTime.start() - start >= duration){
+          options.add(TimeRange.fromStartEnd(start, eventTime.start(), false));
+        }
+        start = eventTime.end();
+      } else if (eventTime.end() > start) {
+        start = eventTime.end();
+      }
+    }
+    if (TimeRange.END_OF_DAY - start >= duration) {
+      options.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
```

Comment:
you should add `/*argument_comment=*/true` for whatever this represents. same above for the `false` case.